### ISSUE TITLE
What are the guidelines for setting loss weights?

### ISSUE THREAD

**Issue by realwecan** [CharlesShang/FastMaskRCNN#123]
@souryuu I have been following your changes to the FastMaskRCNN codebase. Just wondering if you have had any progress/success with the latest mask-visualization branch?

Particularly, I am getting a bit confused about setting the loss weights - it seems they have been changed quite a bit... Can you elaborate a bit on why do you make these changes? These loss weights are rpn_box_lw, rpn_cls_lw, refined_box_lw, refined_cls_lw, and mask_lw.

Thanks!

**Comment by souryuu** [comment]
The lastest version of mask_visualization branch should be able to train the masks of some of the classes with good quality. However, the following problems are still unsolved.

1. Mask is noisy for classes with small number of data 
2. Mask quality at object's edge is still not good. Even groundtruth data of masks are still not perfectly fit with the object's edge. It seems that 28*28 mask is too coarse when roi is large. However, the author did not mention anything about this.
3. There are still a lot of class misclassification from rcnn
4. Testing by setting is_training to False does not work. Bbox and mask mostly disappear. I guess this issue comes from batch normalization settings.
5. As from the paper, during testing mask rois should come from rcnn predicted boxes. [done]
6. Training stop after 81962 iteration.
7. Memory issues.

For loss weights, I manually tuned them because I speculated that losses from mask and rcnn boxes were not appropriately contributed to train shared CNN features. For example, if the losses are 0.01/0.1/0.0001/0.01/0.001 I will set the loss weights to 10/1/1000/10/100.  However, you have to carefully set the learning weight as well.

**Comment by souryuu** [comment]
I just updated the branch. You can now test by using test.py. Mask network will get predicted bounding boxes from rcnn as inputs.
![test_est_535](https://user-images.githubusercontent.com/17999216/28805781-51063198-76a7-11e7-9d5a-438e269651ff.jpg)

I also fixed some mask losses. This should improve mask quality.

**Comment by realwecan** [comment]
@souryuu thank you so much for that, I will give it a go now.

**Comment by Sharathnasa** [comment]
@souryuu even in the test script, you are reading all the details of the image like ground truths, masks from the tfrecords and just displaying on the image right? It's not predicting and generating those.

Even i had written the test script to do the same, where i used to used fed single image without gt and mask info. The predicted outputs was not proper.

Please let me know if my understanding is correct or not.

Regards,
Sharath

**Comment by realwecan** [comment]
@Sharathnasa I don't think this is the case, you can simply set ground truths to None and it will not affect the outputs. However let's hear what @souryuu may have to say on this.

**Comment by souryuu** [comment]
I checked the code. It did not use ground truth even though gt was fed. Anyway, I just rewrote the code to make sure that all the results purely come from the prediction of mask rcnn.

**Comment by realwecan** [comment]
Hi @souryuu , thanks for the prompt reply once again!

I am looking at the issues you've listed, and was wondering if you could shed some light on the following issues:

4. Testing by setting is_training to False does not work. Bbox and mask mostly disappear. I guess this issue comes from batch normalization settings.

--- I don't quite understand this, it seems that in test.py you have already set is_training to False?

7. Memory issues.

--- I am facing a similar issue where the training process will apparently consume a large amount of RAM (system RAM, not GPU RAM) and the program/computer hangs afterwards. Are you mentioning a similar memory issue here? In addition, it this the reason for your training to currently stop at 81962 iterations?

Thanks again for the amazing work!

**Comment by souryuu** [comment]
@realwecan 
I set is_training to False only in some parts of the code. If you look carefully, all the "is_training" that involve with batch normalization are still True. This means BN parameters will change during the test.

I am not sure about the memory issue. The process took a lot of resource than it should be from what i see by "top"
 PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND    
18481 root      20   0 33.955g 4.946g 2.282g S 161.3 15.8  91:40.41 python 

I also have no idea why it always stop at 81962. However, It is fine for me right now since I frequently monitor the training.

**Comment by realwecan** [comment]
@souryuu Thanks for explaining the BN issue for me. Do you mean for example the "_extra_conv_arg_scope_with_bn" function in pyramid_network.py? As you pointed out, I can see that we have is_training=True now, in batch_norm_params. Does this mean the BN parameters will change in testing?

I can confirm that I have the same memory issue here. To partially alleviate (but not  eliminate) the issue, you may try reducing the capacity of the data_queue (I set that to 2 instead of 32 at the moment, and min_after_dequeue=1), and the resident memory usage will be reduced a bit. That may not be the root cause of the problem, though.

**Comment by souryuu** [comment]
Thanks for the info on memory issue.

Yes, BN parameters will change in testing.

**Comment by realwecan** [comment]
@souryuu Just here to report my results with your latest code. It seems that the model has been learning something, but still not satisfactory. In particular, it seems that we have a lot of false alarm detections. I was wondering if we would firstly train the RPN (and possibly Faster RCNN), and then with good object proposals (and scores) at hand, we learn masks as the next step. This resembles the multistage learning process used in Faster RCNN, and perhaps would give us better prediction results.

**Comment by souryuu** [comment]
My results are different from you. I found that rpn and rcnn produce less boxes than it should and the boxes does not precisely fit with objects (iou is at its minimum - 0.7), but I agree with you that multistage learning may give better results.

**Comment by insikk** [comment]
@souryuu Thank you for your loss weights suggestion. I am still curious about the behind reasoning. 

You wrote
> For loss weights, I manually tuned them because I speculated that losses from mask and rcnn boxes were not appropriately contributed to train shared CNN features. For example, if the losses are 0.01/0.1/0.0001/0.01/0.001 I will set the loss weights to 10/1/1000/10/100. However, you have to carefully set the learning weight as well.

What is the intuition to make all weighted losses have similar scale? I agree with your idea that both weighted refine_bbox and weighted refine_class losses should have similar scale. 

What i don't agree is for mask loss. Predicted mask depends on bbox, and class prediction. If bbox and class prediction are wrong, maks loss come from it is nothing but noise. So weighted mask loss should contribute less then weighted refine_bbox and weighted refine_class loss. It would train network first predict bbox and class well, then learn to predict mask. Or, the whole network weight might oscillate somewhere not good. This is my feeling. 

I have no concrete justification for my idea. @souryuu If you can guide us any evidence and literature to learn about this weighting things. It would be really helpful. 

Thank you for your suggestion again. 
BTW, does any one have successfully reproduced the mAP score on Mask R-CNN paper?
